sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/AnalysisContextExtensions.cs

```diff
@@ -18,75 +18,63 @@
+        public static SyntaxTree GetSyntaxTree(this SemanticModelAnalysisContext context) =>
+            context.SemanticModel.SyntaxTree;
 
-        public static void ReportDiagnosticWhenActive(this SyntaxNodeAnalysisContext context, Diagnostic diagnostic)
-        {
-            InternalReportDiagnostic(context.Node.SyntaxTree, diagnostic, d => context.ReportDiagnostic(d));
-        }
-
-        public static void ReportDiagnosticWhenActive(this SyntaxTreeAnalysisContext context, Diagnostic diagnostic)
-        {
-            InternalReportDiagnostic(context.Tree, diagnostic, d => context.ReportDiagnostic(d));
-        }
-
-        public static void ReportDiagnosticWhenActive(this CompilationAnalysisContext context, Diagnostic diagnostic)
-        {
-            InternalReportDiagnostic(context.Compilation.SyntaxTrees.FirstOrDefault(), diagnostic,
-                d => context.ReportDiagnostic(d));
-        }
 
-        public static void ReportDiagnosticWhenActive(this SymbolAnalysisContext context, Diagnostic diagnostic)
-        {
-            InternalReportDiagnostic(context.Symbol.Locations.FirstOrDefault(l => l.SourceTree != null)?.SourceTree, diagnostic,
-                d => context.ReportDiagnostic(d));
-        }
-
-        public static void ReportDiagnosticWhenActive(this CodeBlockAnalysisContext context, Diagnostic diagnostic)
-        {
-            InternalReportDiagnostic(context.CodeBlock.SyntaxTree, diagnostic, d => context.ReportDiagnostic(d));
-        }
+        public static void ReportDiagnosticWhenActive(this SyntaxNodeAnalysisContext context, Diagnostic diagnostic) =>
+            ReportDiagnostic(new ReportingContext(context, diagnostic));
+        public static void ReportDiagnosticWhenActive(this SyntaxTreeAnalysisContext context, Diagnostic diagnostic) =>
+            ReportDiagnostic(new ReportingContext(context, diagnostic));
+        public static void ReportDiagnosticWhenActive(this CompilationAnalysisContext context, Diagnostic diagnostic) =>
+            ReportDiagnostic(new ReportingContext(context, diagnostic));
+        public static void ReportDiagnosticWhenActive(this SymbolAnalysisContext context, Diagnostic diagnostic) =>
+            ReportDiagnostic(new ReportingContext(context, diagnostic));
+        public static void ReportDiagnosticWhenActive(this CodeBlockAnalysisContext context, Diagnostic diagnostic) =>
+            ReportDiagnostic(new ReportingContext(context, diagnostic));
 
-        private static void InternalReportDiagnostic(SyntaxTree tree, Diagnostic diagnostic, Action<Diagnostic> report)
+        private static void ReportDiagnostic(ReportingContext reportingContext)
         {
-            if (SonarAnalysisContext.IsAnalysisDisabled(tree, new[] { diagnostic.Descriptor }) ||
-                !SonarAnalysisContext.ShouldDiagnosticBeReported(tree, diagnostic))
+            // This is the new way SonarLint will handle how and what to report...

```

Comment:
Could assert that only one of the delegates is set.

Reply:
I actually can't because it is not null by default and equality is not working on `Func`. I will try to update the behavior to be null by default instead to equal to a func always returning true.